File: internal/proxy/handler.go
package proxy

import (
	"bytes"
	"io"
	"log"
	"net/http"

	"github.com/google/uuid"

	"github.com/ChainSafe/docker-socket-policy/internal/audit"
	"github.com/ChainSafe/docker-socket-policy/internal/middleware"
	"github.com/ChainSafe/docker-socket-policy/internal/types"
)

// Handler is the HTTP handler for the Docker socket proxy.
type Handler struct {
	router    *Router
	chain     *middleware.Chain
	auditLog  *audit.Logger
	transport *ReverseProxy
}

// NewHandler creates a new proxy handler.
func NewHandler(router *Router, chain *middleware.Chain, auditLog *audit.Logger, transport *ReverseProxy) *Handler {
	return &Handler{
		router:    router,
		chain:     chain,
		auditLog:  auditLog,
		transport: transport,
	}
}

// ServeHTTP handles an incoming HTTP request.
func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	requestID := uuid.New().String()

	body, err := io.ReadAll(r.Body)
	if err != nil {
		log.Printf("[%s] Failed to read body: %v", requestID, err)
		http.Error(w, "failed to read request body", http.StatusBadRequest)
		return
	}
	r.Body.Close()

	route := h.router.Route(r.Method, r.URL.Path, body)

	extra := map[string]interface{}{
		"request_id": requestID,
	}
	if route.Service != "" {
		extra["service"] = route.Service
	}
	if route.Image != "" {
		extra["image"] = route.Image
	}

	// If the route action is deny, return immediately
	if route.Action == types.ActionDeny {
		log.Printf("[%s] DENIED %s %s: %s", requestID, r.Method, r.RequestURI, route.DenyMsg)
		h.auditLog.Deny(r.Method, r.RequestURI, route.DenyMsg, extra)
		http.Error(w, route.DenyMsg, http.StatusForbidden)
		return
	}

	result := h.chain.Execute(r, route)

	if !result.Allowed {
		log.Printf("[%s] DENIED %s %s: %s", requestID, r.Method, r.RequestURI, result.Reason)
		h.auditLog.Deny(r.Method, r.RequestURI, result.Reason, extra)
		http.Error(w, result.Reason, http.StatusForbidden)
		return
	}

	if result.ModifiedBody != nil {
		r.Body = io.NopCloser(bytes.NewReader(result.ModifiedBody))
		r.ContentLength = int64(len(result.ModifiedBody))
	} else {
		r.Body = io.NopCloser(bytes.NewReader(body))
	}

	log.Printf("[%s] ALLOWED %s %s: %s", requestID, r.Method, r.RequestURI, result.Reason)
	h.auditLog.Allow(r.Method, r.RequestURI, result.Reason, extra)

	h.transport.ServeHTTP(w, r)
}

File: ts/src/handler.ts
import { randomBytes } from "node:crypto";
import type { IncomingMessage, ServerResponse } from "node:http";
import { AuditLogger } from "./audit.js";
import { Chain } from "./middleware.js";
import { Action, Router } from "./proxy.js";
import { Transport } from "./transport.js";

export class Handler {
  constructor(
    private router: Router,
    private chain: Chain,
    private audit: AuditLogger,
    private transport: Transport,
  ) {}

  async handle(req: IncomingMessage, res: ServerResponse): Promise<void> {
    const requestId = randomBytes(8).toString("hex");
    const method = req.method ?? "GET";
    const path = req.url ?? "/";

    const bodyBytes = await readBody(req);
    let bodyJSON: Record<string, unknown> | undefined;

    if (bodyBytes.length > 0) {
      try {
        bodyJSON = JSON.parse(bodyBytes.toString("utf-8"));
      } catch {
        // Non-JSON body — pass through for read-only endpoints
      }
    }

    const route = this.router.route(method, path, bodyJSON);

    if (route.action === Action.Deny) {
      const msg = route.denyMsg ?? "denied";
      console.warn(`[${requestId}] DENIED ${method} ${path}: ${msg}`);
      this.audit.deny(method, path, msg);
      res.writeHead(403);
      res.end(msg);
      return;
    }

    if (route.action === Action.CreateContainer && route.policy && route.body) {
      const result = this.chain.execute(method, path, route.policy, route.body);
      if (!result.allowed) {
        console.warn(`[${requestId}] DENIED ${method} ${path}: ${result.reason}`);
        this.audit.deny(method, path, result.reason);
        res.writeHead(403);
        res.end(result.reason);
        return;
      }

      console.log(`[${requestId}] ALLOWED ${method} ${path}`);
      this.audit.allow(method, path);
      await this.transport.forward(req, res, result.modifiedBody);
      return;
    }

    console.log(`[${requestId}] ALLOWED ${method} ${path}`);
    this.audit.allow(method, path);
    await this.transport.forward(req, res, bodyBytes.length > 0 ? bodyBytes : undefined);
  }
}

function readBody(req: IncomingMessage): Promise<Buffer> {
  return new Promise((resolve) => {
    const chunks: Buffer[] = [];
    req.on("data", (chunk: Buffer) => chunks.push(chunk));
    req.on("end", () => resolve(Buffer.concat(chunks)));
  });
}
